libraries/ms-common/inc/spi.h

```diff
@@ -0,0 +1,80 @@
+// HAL SPI
+
+#pragma once
+
+#include <stdint.h>
+
+#include "gpio.h"
+#include "status.h"
+
+// For setting clock polarity
+typedef enum {
+  SPI_CPOLARITY_LOW = 0,
+  SPI_CPOLARITY_HIGH,
+} SPICpol;
+
+// For setting clock phase
+typedef enum {
+  SPI_CPHASE_1EDGE = 0,
+  SPI_CPHASE_2EDGE,
+} SPICpha;
+
+// For setting baud rate prescaler
+typedef enum {
+  SPI_BAUDRATE_2 = 0,
+  SPI_BAUDRATE_4,
+  SPI_BAUDRATE_8,
+  SPI_BAUDRATE_16,
+  SPI_BAUDRATE_32,
+  SPI_BAUDRATE_64,
+  SPI_BAUDRATE_128,
+  SPI_BAUDRATE_256,
+} SPIBaudRate;
+
+// For setting which bit is shifted out first
+typedef enum {
+  SPI_FIRSTBIT_MSB = 0,
+  SPI_FIRSTBIT_LSB,
+} SPIFirstBit;
+
+// SPI init struct
+typedef struct SPISettings {
+  // SPI Settings
+  SPICPolarity polarity;
+  SPICPhase phase;
+  SPIBaudRate baud_rate;
+  SPIFirstBit first_bit;
+
+  // GPIO pin settings
+  GPIOAddress mosi_pin;
+  GPIOAddress miso_pin;
+  GPIOAddress sck_pin;
+  GPIOAddress nss_pin;
+}
+
+// Note: spi_x = 0 for SPI1, spi_x = 1 for SPI2.
```

Comment:
What about devices with only one SPI port? What if they have 4? Also, just for future reference - enums are a much more descriptive choice than `uint8_t`s when you're trying to represent choices.